tests/Routes/QsBatch/QsControllerTest.php

```diff
@@ -57,41 +63,50 @@ public function testImplicitBatchCreation (): void
         QsBatch::factory()->create(['id' => 2, 'done' => 0, 'eventFrom' => 0, 'eventTo' => 0, 'wiki_id' => 99, 'entityIds' => 'a,b']);
         EventPageUpdate::factory()->create(['wiki_id' => 111, 'namespace' => 120, 'title' => 'name']);
 
-        $this->json('GET', $this->route.'/getBatches')
+        $response = $this->json('GET', $this->route.'/getBatches')
             ->assertJsonPath('0.id', 1)
-            ->assertJsonPath('0.done', 1)
+            ->assertJsonPath('0.done', 0)
             ->assertJsonPath('0.wiki.domain', 'test.wikibase.cloud')
             ->assertStatus(200);
 
+        $this->assertNotNull($response->json()[0]['pending_since']);
         $this->assertEquals(QsBatch::query()->count(), 3);
     }
 
     public function testMarkDone (): void {
-        QsBatch::factory()->create(['id' => 1, 'done' => 1, 'eventFrom' => 1, 'eventTo' => 2, 'wiki_id' => 1, 'entityIds' => 'a,b']);
-        QsBatch::factory()->create(['id' => 2, 'done' => 0, 'eventFrom' => 2, 'eventTo' => 3, 'wiki_id' => 1, 'entityIds' => 'c,d']);
-        QsBatch::factory()->create(['id' => 3, 'done' => 0, 'eventFrom' => 3, 'eventTo' => 4, 'wiki_id' => 1, 'entityIds' => 'e,f']);
-        QsBatch::factory()->create(['id' => 4, 'done' => 0, 'eventFrom' => 4, 'eventTo' => 5, 'wiki_id' => 6, 'entityIds' => 'g,h']);
+        QsBatch::factory()->create(['pending_since' => Carbon::now()->subSeconds(1), 'id' => 1, 'done' => 1, 'eventFrom' => 1, 'eventTo' => 2, 'wiki_id' => 1, 'entityIds' => 'a,b']);
+        QsBatch::factory()->create(['pending_since' => Carbon::now()->subSeconds(2), 'id' => 2, 'done' => 0, 'eventFrom' => 2, 'eventTo' => 3, 'wiki_id' => 1, 'entityIds' => 'c,d']);
+        QsBatch::factory()->create(['pending_since' => Carbon::now()->subSeconds(3), 'id' => 3, 'done' => 0, 'eventFrom' => 3, 'eventTo' => 4, 'wiki_id' => 1, 'entityIds' => 'e,f']);
+        QsBatch::factory()->create(['pending_since' => Carbon::now()->subSeconds(4), 'id' => 4, 'done' => 0, 'eventFrom' => 4, 'eventTo' => 5, 'wiki_id' => 6, 'entityIds' => 'g,h']);
 
         $this->json('POST', $this->route.'/markDone', ['batches' => '2,3'])
             ->assertStatus(200);
 
         $this->assertEquals(QsBatch::where('id', 1)->first()->done, 1);
+        $this->assertNotNull(QsBatch::where('id', 1)->first()->pending_since);
         $this->assertEquals(QsBatch::where('id', 2)->first()->done, 1);
+        $this->assertNull(QsBatch::where('id', 2)->first()->pending_since);
         $this->assertEquals(QsBatch::where('id', 3)->first()->done, 1);
+        $this->assertNull(QsBatch::where('id', 3)->first()->pending_since);
         $this->assertEquals(QsBatch::where('id', 4)->first()->done, 0);
+        $this->assertNotNull(QsBatch::where('id', 4)->first()->pending_since);
     }
     public function testMarkFailed (): void {
```

Comment:
This is now terribly named: "marking failed" does not set `failed => true` but just removes the `pending_since` and resets potential `done` values. Should this maybe be called `markNotDone` now?

Asking because the change would span multiple repos.

Reply:
Not sure if the renaming is necessary but I certainly need to ask for clarity now: so a QS batch considered `failed` would mean `batch did not call back as "done" after N requeueing attempts`, similarly to how the laravel job queue works?

Reply:
Yes, this is exactly how it works.